### PR TITLE
Devcontainer

### PR DESCRIPTION
**Pull request by ludeeus**
## Proposed change

Adds: 
- devcontainer (That includes the runtime needed for developing the frontend and running core)
- New tasks: 
  - "Develop Supervisor panel" to develop the Supervisor panel
  - "Run HA Core in devcontainer" to start HA Core
  - "Run HA Core for Supervisor in devcontainer" to start HA with requested input values for a remote proxy add-on

With both the new run tasks a base config directory will be created.

This is compatible with GitHub's codespaces, but the time from start to onboarding takes  ~15min, so stick with local use :see_no_evil: 

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+ENV \
+  DEBIAN_FRONTEND=noninteractive \
+  DEVCONTAINER=true \
+  PATH=$PATH:./node_modules/.bin
+
+# Install nvm
+RUN \
+  su vscode -c \
+    "source /usr/local/share/nvm/nvm.sh && nvm install lts/* 2>&1"
```
Comment (balloob) [inline .devcontainer/Dockerfile:12]: We should install the version mentioned in `.nvmrc`

File: .devcontainer/example_configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+default_config:
```
Comment (balloob) [inline .devcontainer/example_configuration.yaml:1]: Do we prefer this or do we want something like this:

https://github.com/home-assistant/core/blob/dev/script/setup#L19-L26

File: .vscode/tasks.json [hunk 3]
```diff
@@ -37,6 +37,40 @@
         "instanceLimit": 1
       }
     },
+    {
+      "label": "Develop Supervisor panel",
```
Comment (balloob) [inline .vscode/tasks.json:41]: Oops I forgot this one, my bad!

File: .devcontainer/example_configuration.yaml (hunk 2 above)
Comment (ludeeus) [reply .devcontainer/example_configuration.yaml:1]: Sure, I can do that, I already inject some other stuff in there as well

File: .vscode/tasks.json [hunk 4]
```diff
@@ -133,5 +167,45 @@
...
+      },
+      "problemMatcher": [],
+      "runOptions": {
+        "instanceLimit": 1
+      }
+    },
+    {
+      "label": "Run HA Core for Supervisor in devcontainer",
+      "type": "shell",
+      "command": "HASSIO=${input:supervisorHost} HASSIO_TOKEN=${input:supervisorToken} script/core",
```
Comment (balloob) [inline .vscode/tasks.json:187]: Inputs are amazing.

**Review by balloob — APPROVED** [review]

File: script/core [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  echo "Installing Home Asstant core from dev."
+  python3 -m pip install --upgrade \
+    colorlog \
+    git+git://github.com/home-assistant/home-assistant.git@dev
+fi
+
+if [ ! -d "${WORKSPACE}/config" ]; then
+  echo "Creating default configuration."
+  mkdir -p "${WORKSPACE}/config";
+  echo "default_config:
```
Comment (balloob) [inline script/core:28]: I actually meant with my comment, shouldn't we use the `ensure_config` check and then append `demo:` + `logger:` ?

Comment (balloob) [reply script/core:28]: That way we can also drop the `touch` at the end of this file.

Comment (ludeeus) [reply script/core:28]: Ah!
Did not know of that script (looks like it's broken (at least on python 3.9))

File: .vscode/tasks.json [hunk 6]
```diff
@@ -37,6 +37,40 @@
...
+          "activeOnStart": true,
+          "beginsPattern": "Changes detected. Starting compilation",
+          "endsPattern": "Build done @"
+        }
+      },
+      "isBackground": true,
+      "group": {
+        "kind": "build",
+        "isDefault": true
+      },
```
Comment (balloob) [inline .vscode/tasks.json:69]:
```suggestion
      "group": "build",
```

Comment (balloob) [reply .vscode/tasks.json:69]: App is the default.

File: script/core [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  echo "
+# frontend:
+#   development_repo: ${WORKSPACE}
+hassio:
+  development_repo: ${WORKSPACE}" >> "${WORKSPACE}/config/configuration.yaml"
+  else
+  echo "
+frontend:
+  development_repo: ${WORKSPACE}
+# hassio:
```
Comment (balloob) [inline script/core:48]: Let's add new lines or else one might mistake `hassio` to be a child underneath `frontend`
```suggestion

hassio:
  development_repo: ${WORKSPACE}" >> "${WORKSPACE}/config/configuration.yaml"
  else
  echo "
frontend:
  development_repo: ${WORKSPACE}

# hassio:
```

**Pull request merged**